/*
 * Copyright 2021 Google LLC.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

syntax = "proto2";

package yggdrasil_decision_forests.model.distributed_decision_tree.dataset_cache.proto;

import "yggdrasil_decision_forests/dataset/data_spec.proto";

// Configuration for the creation of a cache.
message CreateDatasetCacheConfig {
  // Indicative size of a file in the index cache, expressed in bytes.
  optional int64 index_cache_file_size_bytes = 1 [default = 20000000];  // 20MB

  // Optional index of the label column in the dataspec.
  optional int32 label_column_idx = 2;

  // Optional index of the weight column in the dataspec. Not set if the
  // training is non-weighted.
  optional int32 weight_column_idx = 3;

  // If true, and if "weight_column_idx" is set, the cache does not includes
  // examples with weight = 0.
  optional bool remove_zero_weighted_examples = 4 [default = true];

  // Maximum number of unique value of a numerical feature to allow its
  // pre-discretization. In case of large datasets, discretized numerical
  // features with a small number of unique values are more efficient to learn
  // than classical / non-discretized numerical features.
  //
  // This parameter does not impact the final model. However, it can speed-up or
  // slown the training.
  optional int64 max_unique_values_for_discretized_numerical = 5
      [default = 16000];

  // If false, only the numerical column safisfying
  // "max_unique_values_for_discretized_numerical" will be discretized. If true,
  // all the numerical columns will be discretized. Columns with more than
  // "max_unique_values_for_discretized_numerical" unique values will be
  // approximated with "max_unique_values_for_discretized_numerical" bins.
  //
  // This parameter will impact the model training.
  optional bool force_numerical_discretization = 6 [default = false];
}

// Welcome message of the worker.
message WorkerWelcome {
  // No welcome data yet.
}

// Request message of the workers.
message WorkerRequest {
  // Each of the following actions are made available buy a function of the same
  // name in "column_cache.h".
  oneof type {
    // Separate the columns of a dataset into individual files (per column and
    // per shards).
    SeparateDatasetColumns separate_dataset_columns = 1;
  }

  message SeparateDatasetColumns {
    // Part to dataset (or subset of dataset).
    optional string dataset_path = 1;

    // Output directory.
    optional string output_directory = 2;

    // Columns to exact. The other columns are ignored.
    repeated int32 columns = 3;
    optional dataset.proto.DataSpecification dataspec = 4;

    // Corresponding shard index in the feature cache.
    optional int32 shard_idx = 5;

    // If set, index of a numerical column. The example with zero value (for
    // this column) are removed from the cache.
    optional int32 column_idx_remove_example_with_zero = 6;

    // Number of shards for the feature.
    optional int32 num_shards = 7;
  }
}

// Result message of the worker.
message WorkerResult {
  oneof type {
    SeparateDatasetColumns separate_dataset_columns = 1;
  }

  message SeparateDatasetColumns {
    // optional string output_directory = 1;
    optional int32 shard_idx = 2;
    optional int64 num_examples = 3;
  }
}

// Information relative to one shard in the feature cache.
message ShardMetadata {
  // Number of example in the shard.
  optional int64 num_examples = 1;
}

// Information relative to one pre-sorted column in the index cache.
message SortedColumnMetadata {
  optional CacheMetadata.NumericalColumn metadata = 1;
}

// Metadata relative to the entire cache.
message CacheMetadata {
  // Number of examples in the entire cache.
  optional int64 num_examples = 1;

  // Number of shards i.e. the data of each features is divided into
  // "num_shards" files.
  optional int32 num_shards_in_feature_cache = 2;

  // Number of shards for the index cache.
  optional int32 num_shards_in_index_cache = 3;

  // Information about the columns.
  repeated Column columns = 4;

  // Index of the label column in the dataspec.
  optional int32 label_column_idx = 5;

  // Index of the weight column in the dataspec. Not set if the training is
  // non-weighted.
  optional int32 weight_column_idx = 6;

  message Column {
    // Is the column available in the cache.
    optional bool available = 1 [default = false];

    // Type-specific column information.
    oneof type {
      NumericalColumn numerical = 2;
      CategoricalColumn categorical = 3;
      BooleanColumn boolean = 4;
    }
  }

  message NumericalColumn {
    optional float replacement_missing_value = 1;
    optional int64 num_unique_values = 2;
    optional bool discretized = 3;
    optional int32 discretized_replacement_missing_value = 4;
    optional int64 num_discretized_shards = 5;
    optional int32 num_discretized_values = 6;
  }

  message CategoricalColumn {
    optional int64 num_values = 1;
    optional int32 replacement_missing_value = 2;
  }

  message BooleanColumn {
    optional bool replacement_missing_value = 2;
  }
}

message DatasetCacheReaderOptions {
  // Indices of the features accessible in reading operations.
  // If empty, all the features are available.
  repeated int32 features = 1 [packed = true];

  // Number of values to read at each reading call.
  optional int32 reading_buffer = 2 [default = 2000];

  // Load an read the cache from memory, or read the cache from disk.
  optional bool load_cache_in_memory = 3 [default = true];
}
